### PR TITLE
Implement token revocation

### PR DESCRIPTION
**Pull request by jameswestman**
For tokens with a "jti" field, query the database to see if the token has been revoked. Also, log the time it was last used, to help debug and/or recognize suspicious activity.

There are two new endpoints, to list tokens (so the backend can show the last used time and whether a token is revoked) and to revoke tokens. The idea is for the backend to keep its own record of which tokens belong to which users, and invoke these endpoints on their behalf with its own token.

I've also added a config option to allow tokens to be prefixed. This makes it easier to recognize tokens that have been accidentally published.

### CODE REVIEW

File: src/tokens.rs [hunk 1]
```diff
@@ -275,76 +291,114 @@ where
...
+            service: Rc::new(Mutex::new(service)),
             inner: self.0.clone(),
         })
     }
 }
 
 /// TokenParser middleware
 pub struct TokenParserMiddleware<S> {
-    service: S,
+    service: Rc<Mutex<S>>,
```
Comment (A6GibKm) [inline src/tokens.rs:302]: Any reason to use `Mutex` instead of `RefCell`/`Cell`/`OnceCell`?

File: src/api/tokens.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#[derive(Deserialize)]
+pub struct TokenArgs {
+    token_ids: Vec<String>,
+}
+
+pub fn get_tokens(
+    args: Json<TokenArgs>,
+    db: Data<Db>,
+    req: HttpRequest,
+) -> impl Future<Item = HttpResponse, Error = ApiError> {
```
Comment (A6GibKm) [inline src/api/tokens.rs:20]: Any reason to not use the `async` keyword? It does more than just being eye candy.

File: src/api/tokens.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+use crate::db::Db;
+use crate::errors::ApiError;
+use crate::tokens::{ClaimsScope, ClaimsValidator};
+
+#[derive(Deserialize)]
+pub struct TokenArgs {
+    token_ids: Vec<String>,
+}
+
+pub fn get_tokens(
```
Comment (A6GibKm) [inline src/api/tokens.rs:16]: As Jamie pointed out, this is just as async as the other function, why have two?

File: src/api/tokens.rs (hunk 2 above)
Comment (jameswestman) [reply src/api/tokens.rs:20]: We are using a very old version of Actix that uses futures 0.1, so the extra functions convert from std futures to the old ones with `.compat()`.

File: src/tokens.rs (hunk 1 above)
Comment (jameswestman) [reply src/tokens.rs:302]: RefCell should work. Fixed.

**Comment by bilelmoussaoui** [comment]
I would appreciate if we could land https://github.com/flatpak/flat-manager/pull/92 so we don't have to keep adding more "old" way of doing things from diesel side.

**Pull request merged**